src/ring_theory/witt_vector/mul_p.lean

```diff
@@ -0,0 +1,84 @@
+/-
+Copyright (c) 2020 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin
+-/
+
+import ring_theory.witt_vector.is_poly
+
+/-!
+## Multiplication by `n` in the ring of Witt vectors
+
+In this file we show that multiplication by `n` in the ring of Witt vectors
+is a polynomial function. We then use this fact to show that the composition of Frobenius
+and Verschiebung is equal to multiplication by `p`.
+
+### Main declarations
+
+* `mul_n_is_poly`: multiplication by `n` is a polynomial function
+
+-/
+
+namespace witt_vector
+
+variables {p : ℕ} {R : Type*} [hp : fact p.prime] [comm_ring R]
+local notation `𝕎` := witt_vector p -- type as `\bbW`
+
+local attribute [semireducible] witt_vector
+
+open mv_polynomial
+noncomputable theory
+
+include hp
+
+variable (p)
+
+/-- `witt_mul_n p n` is the family of polynomials that computes
+the coefficients of `x * n` in terms of
+the coefficients of the Witt vector `x`. -/
+noncomputable
+def witt_mul_n : ℕ → ℕ → mv_polynomial ℕ ℤ
+| 0     := λ k, 0
+| (n+1) := λ k, bind₁ (function.uncurry $ ![(witt_mul_n n), X]) (witt_add p k)
+
+variable {p}
+
+lemma mul_n_coeff (n : ℕ) (x : 𝕎 R) (k : ℕ) :
+  (x * n).coeff k = aeval (λ i, x.coeff i) (witt_mul_n p n k) :=
+begin
+  induction n with n ih generalizing k,
+  { simp only [nat.nat_zero_eq_zero, nat.cast_zero, mul_zero,
+      zero_coeff, witt_mul_n, alg_hom.map_zero], },
+  { rw [witt_mul_n],
+    simp only [nat.succ_eq_add_one, mul_add, mul_one, nat.cast_add, nat.cast_one],
+    rw [aeval_bind₁, add_coeff],
+    apply eval₂_hom_congr (ring_hom.ext_int _ _) _ rfl,
+    ext1 ⟨b, i⟩,
+    fin_cases b;
+    simp only [function.uncurry, ih, aeval_X,
+      matrix.cons_val_zero, matrix.head_cons, matrix.cons_val_one], }
```

Comment:
These `simp`s are quite slow: like 1 second total on my machine. Here's an attempt to halve the time:

```suggestion
  { rw [witt_mul_n, nat.succ_eq_add_one, nat.cast_add, nat.cast_one, mul_add, mul_one,
      aeval_bind₁, add_coeff],
    apply eval₂_hom_congr (ring_hom.ext_int _ _) _ rfl,
    ext1 ⟨b, i⟩,
    fin_cases b,
    { simp only [function.uncurry, matrix.cons_val_zero, ih] },
    { simp only [function.uncurry, matrix.cons_val_one, matrix.head_cons, aeval_X] } }
```